### PR TITLE
Reworked fix for #2375

### LINKED ISSUE #2375: New `check_only_running_once()` call failing on macOS 10.13.6; test fails also.
### What were you trying to do?

build and run mu on macOS 10.13.6 using python 3.8.10

### What steps did you take to trigger the issue?

// setup pyenv veritual environment & activate
// set it to use python 3.8.10 (latest 3.8)
// git clone https://github.com/mu-editor/mu.git
// do dev setup `pip install -e ".[dev]"` etc.

`make run`
fails

`make test` fails one test relating to `check_only_running_once()`

Checking out the head -1 commit results in running successfully every time and all tests passing.

### What did you expect to happen?

run successfully, all tests pass

### What actually happened?

`make run`

results in:

> (venv) haikumintthin:mu tyler$ make run
> rm -rf build
> rm -rf dist
> rm -rf .coverage
> rm -rf .eggs
> rm -rf *.egg-info
> rm -rf docs/_build
> rm -rf .pytest_cache
> rm -rf lib
> rm -rf *.mp4
> rm -rf .git/avatar/*
> rm -rf venv-pup
> find . \( -name '*.py[co]' -o -name dropin.cache \) -delete
> find . \( -name '*.bak' -o -name dropin.cache \) -delete
> find . \( -name '*.tgz' -o -name dropin.cache \) -delete
> find . | grep -E "(__pycache__)" | xargs rm -rf
> rm -f ./mu/locale/messages.pot
> rm -f ./mu/wheels/*.zip
> python run.py
> make: *** [run] Error 2
> 

----

`make test` results in:

> ... <snip tons of output>

> ======================================================================================== FAILURES =========================================================================================
> _________________________________________________________________________________ test_only_running_once __________________________________________________________________________________
> 
>     def check_only_running_once():
>         """If the application is already running log the error and exit"""
>         try:
>             with _shared_memory:
> >               _shared_memory.acquire()
> 
> mu/app.py:316: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <mu.app.SharedMemoryMutex object at 0x10c4bbbe0>
> 
>     def acquire(self):
>         if self._shared_memory.attach():
>             pid = struct.unpack("q", self._shared_memory.data()[:8])
> >           raise MutexError("MUTEX: Mu is already running with pid %d" % pid)
> E           mu.app.MutexError: MUTEX: Mu is already running with pid 72090
> 
> mu/app.py:291: MutexError
> 
> During handling of the above exception, another exception occurred:
> 
>     def test_only_running_once():
>         """
>         If we are the first to acquire the application lock we should succeed
>         """
> >       check_only_running_once()
> 
> tests/test_app.py:419: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
>     def check_only_running_once():
>         """If the application is already running log the error and exit"""
>         try:
>             with _shared_memory:
>                 _shared_memory.acquire()
>         except MutexError as exc:
>             [message] = exc.args
>             logging.error(message)
> >           sys.exit(2)
> E           SystemExit: 2
> 
> mu/app.py:320: SystemExit
> ------------------------------------------------------------------------------------ Captured log call ------------------------------------------------------------------------------------
> ERROR    root:app.py:319 MUTEX: Mu is already running with pid 72090
> ================================================================================= short test summary info =================================================================================
> FAILED tests/test_app.py::test_only_running_once - SystemExit: 2
> ======================================================================= 1 failed, 1116 passed, 20 skipped in 30.98s =======================================================================
> make: *** [test] Error 1
> 

----
Oddly, I can get it to run if I run it in the debugger.  It fails but then appears to re-run itself (possibly some aspect of the debugger?).

> (venv) haikumintthin:mu tyler$ python3 -m pdb run.py
> > /Users/tyler/Documents/Work/mu/run.py(2)<module>()
> -> from mu.app import run
> (Pdb) c
> The program exited via sys.exit(). Exit status: 2
> > /Users/tyler/Documents/Work/mu/run.py(2)<module>()
> -> from mu.app import run
> (Pdb) c
> QSharedMemory::lock: already locked
> 

Screen Shot 2022-12-24 at 6 22 54 PM

-----

Looking at the log, it appears that it's somehow getting the shared memory space from the previous successful run.  Like it's not being cleaned up when the process exits and is somehow in a global namespace where it can be found again (?).

I'm going to restart my machine and try again to see what a _definitely not in memory_ situation results in.

### Operating System Version

_No response_

### Mu Version

_No response_

### Other Info

_No response_

### Editor Log

[mu.log](https://github.com/mu-editor/mu/files/10299361/mu.log)

### PR DESCRIPTION
**Pull request by haikusw**
Last fix was _not_ a fix.  Broke multiple copy running test (on macOS at least), despite all tests passing.

### CODE REVIEW

File: tests/test_app.py [hunk 1]
```diff
@@ -431,12 +434,28 @@
...
+    If we run, throw an exception and then run again we should succeed.
+    """
+    setup_logging()  # installs excepthook
+    assert sys.excepthook == excepthook
+    check_only_running_once()  # leaves shared memory attached
+    ex = MutexError("BOOM")
+    with pytest.raises(MutexError):
+        raise ex
+    # shared memory should be cleaned up, so check running once again
+    check_only_running_once()
```
[the remark this answers is not in the document]
Comment (haikusw) [reply tests/test_app.py:460]: The bot is confused, seems to me.

**Comment by haikusw** [comment]
Ok.  Reworked this fix and added at least partial coverage for clearing the shared memory on exceptions. 
 Updated concurrent version tests, and added new test for exception handler clearing of shared memory on exception before exiting.

Basic idea was to add a `SharedMemoryMutex.release()` method that can be called before app exit to clean up the shared memory attachment done by this instance.  Then call that method before calling exit and for exception exits (crash).

Implementation in app.py:
 - add the `release()` method to `SharedMemoryMutex`.
 - update the `excepthook` implementation to call `_shared_memory.release()` before calling `sys.exit`.
 - remove the incorrect _shared_memory.detach() call in `__exit__`.
 - update the end of `run():` to call `app_exec_()`, save the exit_status, call `_shared_memory.release()`, and then call `sys.exit` with the saved exit_status.

Then I had to modify the explicit `sys.exit(0)` in logic.py on line 1431 so that it wouldn't short circuit exit the app and cause `app_exec_()` to never return and thus our exit clean up of the shared memory not get called.

Changing logic.py line 1431 from `sys.exit(0)` to `QtCore.QCoreApplication.exit(0)` results in `app_exec_()` returning and the cleanup code I added in app.py `run():` to call the new `_shared_memory.release()` method and this appears to fix the issue for real for the clean no-error exit path.

I did have to rework test_logic.py's `test_quit_calls_sys_exit(mocked_session)` to expect the right thing instead of `sys.exit` explicitly. 

Then I modified the existing tests in app.py:
 - `test_only_running_once()` to clean up after itself by calling `_shared_memory.release()` explicitly since it is a focused test that doesn't run the app event loop.
 - similarly `test_running_twice()` had to be modified to also have the first run process clean up after itself when it exits.

Finally, I added a new test to test_app.py test the exception shared memory cleanup handling, `test_running_twice_after_exception()`.

----
Caveats - there may be other places where `sys.exit` is being called that short circuits the `app_exec()` event loop and prevents the shared memory cleanup code from getting called.   That will be no worse than the master was before I tried to fix this the first time, but it would be good to track those down and convert them to `QtCore.QCoreApplicaton.exit` or `QApplication.exit`, as appropriate for the scope they are in.

----
Sorry this got a bit more complicated, but I believe it's a more complete fix now since it adds at least some of the exception path cleanup whereas previously it did not.
 
It looks like there may be issues on non-macOS platforms that CI is calling up.  I'll have to look at those a bit later since it's now 5am here.

**Comment by haikusw** [comment]
Looks like even though `make check` works here on macOS 10.13 with python 3.8.16, there are other configurations where `test_subprocess_run_invalid_utf8()` is now failing.

I'll have to look at those a bit later since it's now 5am here.

**Comment by haikusw** [comment]
@tjguk As noted above - my last attempt (merged to master) wasn't a good fix - it just disabled the run twice code, effectively (and yet the `test_running_twice()` test still passed?!).

This seems closer to what we want; what do you think?

**Comment by tjguk** [comment]
Thanks for putting the work in here @haikusw -- I'm about to go away for a week so I might not get to comment substantively on your PR before I go. But I will definitely look at it when I get back. And feel to ping me for an update if I don't!

**Comment by haikusw** [comment]
@tjguk Thanks for letting me know.  Hoping to work on this a bit more anyway because it bothers me that the `test_running_twice()` test didn't catch my error.

I hope you have safe and smooth travels and a good week away.  Cheers.

**Comment by haikusw** [comment]
The `test_running_twice()` test did not seem to be doing what it appeared to be trying to do.  It also had timing issues in that the subprocesses could run out of order which would make the test fail sometimes (on this older macOS machine at least).  Rewrote it so that it's reliable on this mac, but looking to see what CI and others see elsewhere.

----

This seems about as good as I can make this without more input.  I do not think this approach is particularly good because any code that is added that calls `sys.exit` is going to circumvent the clean up of the shared memory block and require the user to restart their computer (for macOS at least) to be able to run Mu again.  

I have some thoughts on what could be done to avoid that, but this would require some discussion before the work to implement it.

If the current approach is retained, I think it would be good to let the user know they have to restart to run Mu again in the case where it crashes and `excepthook` isn't called to clean up.

Also, as noted in a commit comment, `setup_logging()` should probably be changed to always install `excepthook` instead of only if `MU_LOG_TO_STDOUT` env var isn't set.

----

Anyway, we'll see what CI says about my latest changes and, when @tjguk gets back from traveling, perhaps we can discuss this more.

**Comment by haikusw** [comment]
Weird that this piece of code I didn't change is now causing a test failure…

> out:     self._shared_memory.data()[:8] = struct.pack("q", os.getpid())

… and only on 2 of 3 PiOS test runners?  🤔

**Comment by haikusw** [comment]
(yep).  

If there was a way to wait until a `Popen`-ed process had completed launching then we could do that instead of the `sleep` calls, but I didn't see a way.  Can wait until it _exits_, and do later in the test, but that's not what we want for the middle where we wait to launch `cmd2`.

**Comment by haikusw** [comment]
The discussion in #276 seem relevant.  In particular, `QLockFile` seems like it might be a better choice than `QSharedMemory` since `QLockFile` will, apparently, auto-cleanup the lock file if it's stale (if you use it correctly).

https://stuff.mit.edu/afs/athena/software/texmaker_v5.0.2/qt57/doc/qtcore/qlockfile.html#details

**Comment by haikusw** [comment]
Hey @ntoll, welcome back.

First thing:  This needs review by someone more experienced with Python, pyQt/Qt.  I am a novice when it comes to these technologies and this is definitely not a novice level thing. While I do have 35ish years experience as a professional developer, it's largely all native sdk Apple platforms work.  So my lack of knowledge with these technologies and the number of people potentially impacted seems like it requires a careful review by those more experienced than I.

I think it’s good (with the caveats / additional work to consider I noted), but…. 


Second thing:  I outlined above some of the limitations in what I’ve done and the general approach that was present which this PR tries to improve (bit brittle, any misses of exception hook (added later or missed calls to sys.exit) will require a restart by users to fix, complex to test, etc), a few more things that could be done to improve it, and asked the question about why you didn’t go with  `QLockFile` approach mentioned in #276.  

I’m not sure it’s worth making the current approach as good as it could be if ultimately `QLockFile` is a better approach.

If you did want to stick with this approach then I’d like to know some manual testing on other platforms indicates it seems to work for everyone.  Tests all pass but… they do for main now also and that doesn’t actually work (this PR fixes the false positive signaling test).

Happy to discuss, answer questions about my changes or additional suggested work, etc, here or elsewhere.

**Comment by haikusw** [comment]
So @carlosperate outlined some important limitations with `QFileLock` in #276 and that clarifies this alternate approach being worth pursuing further to completion (thank you).

Work I think should be done on this PR to complete it:

**Seems Required**

- [ ] Rework global exception handler hook (`excepthook`)
    - [ ] Move installation of it out of `setup_logging()` since it'll be used for more than logging.
    - [ ] Exception handler hook needs to always be installed and not be contingent upon the `MU_LOG_TO_STDOUT` environment variable.
    - [ ] `excepthook` needs some rework to add the exclusion of exception logging if the `MU_LOG_TO_STDOUT` env var is set.

**Would be good**

Strongly recommend this, but isn't *required* I guess.

- [ ] Add an Alert when `check_only_running_once()` says a launched instance has to quit that alerts the user to there being another instance running and that if they are not finding one then they should restart their computer. (yuck, I know, but better than a confused user thinking the app is just broken and crashing on launch).

This seems potentially complex because it sort of "soft runs" the app just to show this alert and exit but should avoid all the other initialization work gets started and where and what might need to be prevented from running, etc.  For example, `settings.init()` is called immediately after `check_only_running_once()` is called...

This complexity of the app startup sequence internals, plus my zero experience with PyQt or Qt and very limited python experience, makes me hesitate to volunteer to implement this particular piece.

**Phase two improvement**

Later improvement to avoid the user having to restart their computer in the case of a crashed app that didn't clean up the shared memory buffer.   This is a proposed mechanism and should be discussed for anything I'm missing/haven't thought of etc:

- [ ] Add a similar "auto-recovery" mode to the shared memory multiple-instance exclusion mechanism that the `QFileLock` has:
    - [ ]  Add a periodic "I'm still here" timestamp field to the shared memory buffer.
    - [ ] Have the registered PID process update that at some TBD interval with the current time.
    - [ ] Enhance `check_only_running_once()` to check this timestamp and to take over the shared memory buffer if too much time had passed.  update that only the registered process id would update and the ability for another process to "take over" the shared memory buffer by overwriting the PID field with its PID and to start updating the timestamp and then to allow the app to run. 
    
I don't think this addition is necessary to ship the fix and could be a second phase of implementing this, but it would be a significant usability improvement in my opinion.

----

My biggest concerns with the shared memory multiple-instance mechanism as implemented are:

a) The user will not understand why the app is "crashing" on launch (from their perspective) and won't know that to do.

b) This approach feels brittle, as noted above.  Any current `sys.exit` calls or ones that are added in the future are going to skip the global exception hook and thus not clean up the shared buffer and put the user in the stuck situation of a "crashing on launch" Mu app.

If those are acceptable, perhaps by adding the alert telling the user what to do, and if there are ways to reduce the risk of calls to `sys.exit` being added later (hook it? that seems iffy, but I'm too much a novice on python to know), then I can work on the additions in the "Required** section above.

Hope that's clear.  Medication is making my brain a little foggy today.

**Comment by haikusw** [comment]
Do either of you know why the exception handler is not installed by `setup_logging()` if "MU_LOG_TO_STDOUT" is set in the environment?

**Comment by haikusw** [comment]
(test failure _seems_ to be the timing of the running twice tests on the slow PiOS testers.  I can't kick the workflow to get it to try again to see if it was just bad luck or if I need to increase those tests times more).

**Comment by carlosperate** [comment]
> Do either of you know why the exception handler is not installed by `setup_logging()` if `MU_LOG_TO_STDOUT` is set in the environment?
> 
> I have changes made (as updated via checkboxes above), but I'm unsure about this detail. TIA.

I'm not 100% sure, but because the goal of setting up the `MU_LOG_TO_STDOUT` env var is to get all the data on the terminal, it also makes sense to get the exception printed in the terminal instead of captured and opened in a browser.

> (test failure _seems_ to be the timing of the running twice tests on the slow PiOS testers. I can't kick the workflow to get it to try again to see if it was just bad luck or if I need to increase those tests times more).

Okay, I've restarted the two failed PiOS jobs.


Thanks again for your work on this feature!

**Comment by haikusw** [comment]
Thanks @carlosperate

> I'm not 100% sure, but because the goal of setting up the MU_LOG_TO_STDOUT env var is to get all the data on the terminal, it also makes sense to get the exception printed in the terminal instead of captured and opened in a browser.

Reading the `logging` documentation, it seems like log handlers added with `addHandler` are _additive_ (?).   If so, as long as setting `MU_LOG_TO_STDOUT` results in the standard out log handler getting added via `addHandler` then having the exception logging to _all_ installed handlers should result in that information still showing up on standard out?


The odd thing is that `StreamHandler()` as setup in `setup_logging()` doesn't specify `stdout` as the arguement, so it's going to go to stderr, if the [documentation is to be believed ](https://docs.python.org/3.8/library/logging.handlers.html?highlight=streamhandler#logging.StreamHandler)(?).  To get it to go to stdout one would have to do something like:

`stdout_handler = logging.StreamHandler(stream=sys.stdout)`

**Comment by haikusw** [comment]
Is there a way to set an environment variable in test.yml (??) that the automated tests can read that applies to only PiOS runs of the tests?  I'd like to be able to have the timeout for the multi-instance tests be shorter for all the fast OSs but slow enough to be reliable on PiOS (which seems really slow to launch Mu).

**Comment by haikusw** [comment]
Ok - added two things:

1. Now [using](https://github.com/mu-editor/mu/pull/2386/commits/e87f6b12a43c923d438f66f6aca227f7ddbf88e4) a dynamic app name for the registration of the multi-instance execution prevention shared memory buffer during automated test running.  This means test failures don't forcing the developer to reboot their computer to clear out the one and only shared memory buffer.  This is enabled by setting the environment variable `MU_TEST_SUPPORT_RANDOM_APP_NAME_EXT` and is set for all automated tests via a test fixture in [conftest.py](https://github.com/mu-editor/mu/pull/2386/commits/e87f6b12a43c923d438f66f6aca227f7ddbf88e4#diff-e52e4ddd58b7ef887ab03c04116e676f6280b824ab7469d5d3080e5cba4f2128R35)
1. [Add an environment variable](https://github.com/mu-editor/mu/pull/2386/commits/29db433e60a2efbef4ff7d279a762420796ec597) `MU_TEST_SUPPORT_SLOW_RUNNER_FLAG` to signal a need to increase process launch waiting times for the multi-instance prevention automated test `test_running_twice()` in test_app.py because PiOS CI instances take a _really_ long time to launch processes.  Simply increasing the timeouts across all test execution targets slowed automated testing everywhere despite only being necessary for PiOS targets.  Environment flag is set in GitHub action file test.yml for the `test-pios:` target/runs only, though it could be set on the command line if someone had a slow machine they were testing on.
_NOTE:_ if this test fails on _non PiOS_ machines because I dialed the "normal" test case wait times down too far, they can be increased in `test_running_twice()` as documented [here in app_test.py](https://github.com/mu-editor/mu/pull/2386/commits/29db433e60a2efbef4ff7d279a762420796ec597#diff-a67cb1853203a6f1956991a9d9881d231c4d43557f5baffd45cc672a87e41cc6R452).

The only outstanding "before merge" issue I'm aware of is the conversation around the `excepthook` always installed  change [as discussed above](https://github.com/mu-editor/mu/pull/2386#issuecomment-1385030909) and the follow-on comments.

The other thing I'd _like_ to do is document this multi-instance prevention strategy & implementation somewhere (and these new environment variables I've added), but I haven't had a chance to go figure out _where_ that belongs (pointers gladly accepted!).

**Comment by tjguk** [comment]
Thanks for pinging @haikusw -- I'm afraid I managed to catch pneumonia while I was away and ended up in hospital for a few days. So not only have I been playing catch-up ever since: I'm also running a little bit under 100% still. I shall definitely take a look at what you've done; and thank you so much for persevering in this.

**Comment by haikusw** [comment]
> Thanks for pinging @haikusw -- I'm afraid I managed to catch pneumonia while I was away and ended up in hospital for a few days. So not only have I been playing catch-up ever since: I'm also running a little bit under 100% still. I shall definitely take a look at what you've done; and thank you so much for persevering in this.

Yikes!  No worries at all.  Getting fully well is priority one.  Pneumonia can take a while to fully kick.  Best of good luck to you on a speedy recovery!

**Comment by haikusw** [comment]
Looking at this code more, the lack of result checking in the original `SharedMemoryMutex` `acquire` function is concerning.  Additionally, don't we have to `lock` the shared memory after `attach` fails and we call `create` on line 304 (and don't we have to check for that result before writing to `data()`?).  The call to `lock` on `__enter__` isn't going to work because it hasn't been `create`d yet, seems like.

I started looking at this after reading this in [the log](https://github.com/mu-editor/mu/actions/runs/3951760857/jobs/6766014899#step:9:1271) for the failed test (which likely just failed due to not enough time, but I haven't confirmed that):

> [1271](https://github.com/mu-editor/mu/actions/runs/3951760857/jobs/6766014899#step:9:1271)

```
       assert result1 == 0
out: E       assert 1 == 0
out: tests/test_app.py:508: AssertionError
out: ----------------------------- Captured stdout call -----------------------------
out: processOneLaunchTime: 2
out: processTwoLaunchAndExitTime: 4
out: process 2 id: 1065
out: process 1 id: 1064
out: ----------------------------- Captured stderr call -----------------------------
out: /usr/lib/python3/dist-packages/requests/__init__.py:91: RequestsDependencyWarning: urllib3 (1.26.14) or chardet (3.0.4) doesn't match a supported version!
out:   RequestsDependencyWarning)
out: ERROR:root:Unrecoverable error
out: Traceback (most recent call last):
out:   File "<string>", line 1, in <module>
out:   File "/home/pi/mu/mu/app.py", line 330, in check_only_running_once
out:     _shared_memory.acquire()
out:   File "/home/pi/mu/mu/app.py", line 305, in acquire
out:     self._shared_memory.data()[:8] = struct.pack("q", os.getpid())
out: TypeError: 'NoneType' object does not support item assignment
out: /usr/lib/python3/dist-packages/requests/__init__.py:91: RequestsDependencyWarning: urllib3 (1.26.14) or chardet (3.0.4) doesn't match a supported version!
out:   RequestsDependencyWarning)
out: Traceback (most recent call last):
out:   File "<string>", line 1, in <module>
out:   File "/home/pi/mu/mu/app.py", line 330, in check_only_running_once
out:     _shared_memory.acquire()
out:   File "/home/pi/mu/mu/app.py", line 305, in acquire
out:     self._shared_memory.data()[:8] = struct.pack("q", os.getpid())
out: TypeError: 'NoneType' object does not support item assignment
out: =============================== warnings summary ===============================
```

> out: TypeError: 'NoneType' object does not support item assignment

got me started digging.  Maybe just my ignorance.

**Comment by haikusw** [comment]
The PiOS tests aren't passing 100% reliably on CI on my branch when I redo them manually.  Going to have to add some more logging to be able to understand what's going on more.  Tests sensitive to timing are such a drag.  May need to rethink how this test is written entirely.  Figure out a way to know 100% that the first process has finished launching and has the shared memory buffer allocated before running the second process.  Or maybe the second process runs but spins testing until the first one has created the memory buffer (as a way to know it's running) and then does it's test...  Or maybe the second process can launch (detached somehow?) the second process?
Something to ponder a bit.  May have to put together a small PyQt stand-alone test app to make it easier to work on efficiently.  Not today though.

**Comment by haikusw** [comment]
I was working on removing the timing issues with the tests and ran into a question:

Are you sure that Popen vs subprocess.run opens the processes in separate process trees?  They seem to have the same process group id when I execute a test program on macOS 10.12.x is why I ask.

Also, are we sure this is an actual requirement for the QSharedMemory exclusion to work?

**Comment by haikusw** [comment]
I'm closing this in favor of #2415 (though I need someone to try that on Windows to make sure it works there).  We can always re-open it if #2415 doesn't work out.

**Pull request closed without merging**